raiden/transfer/architecture.py

```diff
@@ -138,3 +138,23 @@ def dispatch(self, state_change):
         assert all(isinstance(e, Event) for e in events)
 
         return events
+
+    def __eq__(self, other):
+        if isinstance(other, StateManager):
```

Comment:
the problem is, that `__slots__` don't guarantee the existence of an `attribute`, i.e. see this error:

```
    def __eq__(self, other):
        if isinstance(other, TargetState):
            return (
                self.our_address == other.our_address and
                self.from_route == other.from_route and
                self.from_transfer == other.from_transfer and
                self.block_number == other.block_number and
>               self.secret == other.secret and
                self.state == other.state
            )
E           AttributeError: secret

raiden/transfer/mediated_transfer/state.py:169: AttributeError
```

I agree with the "generic" way being complex, but I would argue that adding specific `hasattr` checks for all `__slots__` by name isn't much less complex/difficult to read?

Would you be more comfortable with defining
```
def safe_equal_attribute_value(attribute, self, other):
  if hasattr(self, attribute) and hasattr(other, attribute):
    return self.__getattribute__(attribute) == other.__getattribute__(attribute)
  else:
    return not hasattr(self, attribute) and not hasattr(other, attribute)
```
and calling `safe_equal_attribute_value` on all defined slots specifically?

Or do you have a better suggestion?